File: classes/lista3.scala
// Jakub Czerwiński

// Zadanie 2
def curry3[A, B, C, D] (f:(A, B, C) => D)(x:A)(y:B)(z:C) = f(x, y, z)
def curry3L[A, B, C, D] (f:(A, B, C) => D) : A => B => C => D = x => y => z => f(x, y, z)

def uncurry3[A, B, C, D] (f:A => B => C => D)(x:A, y:B, z:C) = f(x)(y)(z)
def uncurry3L[A, B, C, D] (f:A => B => C => D) : (A, B, C) => D = (x, y, z) => f(x)(y)(z)

def plus (x:Int)(y:Int)(z:Int) : Int = x + y + z
def add (x:Int, y:Int, z:Int) : Int = x + y + z

curry3(add)(1)(2)(3)
curry3L(add)(1)(2)(3)

uncurry3(plus)(1, 2, 3)
uncurry3L(plus)(1, 2, 3)

// Zadanie 3
def mySumProd(xs:List[Int]) =
  xs.foldLeft (0, 1) ((t: (Int, Int), x: Int) => (t._1 + x, t._2 * x))

mySumProd(List(1, 2, 3, 4))
mySumProd(List(0, 1, 2, 3))
mySumProd(List(-1, -2, -3, -4))

// Zadanie 5
// a)
def insertionSort[A](pred: (A, A) => Boolean)(xs: List[A]) : List[A] =
  def insert(pred: (A, A) => Boolean)(x: A)(xs: List[A]) : List[A] =
    xs match
      case Nil => List(x)
      case h::t =>
        if pred(x, h) then x::h::t
        else h::insert(pred)(x)(t)

  xs match
    case Nil => Nil
    case h::t => insert(pred)(h)(insertionSort (pred)(t))

insertionSort ((x: Int, y: Int) => x <= y)(List(5, 2, 1, 4, 3))
insertionSort ((x: (Int, _), y: (Int, _)) => x._1 <= y._1)(List((5, 1), (2, 1), (5, 2), (4, 1), (2, 2)))
insertionSort ((x: (Int, _), y: (Int, _)) => x._1 >= y._1)(List((5, 1), (2, 1), (5, 2), (4, 1), (2, 2)))

// b)
def mergeSort[A](pred: (A, A) => Boolean)(xs: List[A]) : List[A] =
  def halve(xs: List[A]) : (List[A], List[A]) =
    def halveHelper(n: Int)(ys: List[A]) : (List[A], List[A]) =
      if n == 0 then (Nil, ys)
      else ys match
        case Nil => (Nil, Nil)
        case h::t =>
          val (left, right) = halveHelper(n - 1)(t)
          (h::left, right)
    halveHelper(xs.length / 2)(xs)
  
  def merge(pred: (A, A) => Boolean)(xs: List[A])(ys: List[A]) : List[A]=
    (xs, ys) match
      case (Nil, list) => list
      case (list, Nil) => list
      case (hx::tx, hy::ty) =>
        if pred(hx, hy) then hx::merge(pred)(tx)(ys)
        else hy::merge(pred)(xs)(ty)
  
  xs match
    case Nil => Nil
    case List(x) => List(x)
    case h::t =>
      val (left, right) = halve(h::t)
      merge(pred)(mergeSort(pred)(left))(mergeSort(pred)(right))

mergeSort ((x: Int, y: Int) => x <= y)(List(5, 2, 1, 4, 3))
mergeSort ((x: (Int, _), y: (Int, _)) => x._1 <= y._1)(List((5, 1), (2, 1), (5, 2), (4, 1), (2, 2)))
mergeSort ((x: (Int, _), y: (Int, _)) => x._1 >= y._1)(List((5, 1), (2, 1), (5, 2), (4, 1), (2, 2)))

File: classes/lista3.ml
(* Jakub Czerwinski *)

(* Zadanie 2 *)
let curry3 f x y z = f (x, y, z);;
let curry3L = fun f -> fun x -> fun y -> fun z -> f (x, y, z);;

let uncurry3 f (x, y, z) = f x y z;;
let uncurry3L = fun f -> fun (x, y, z) -> f x y z;;

let plus x y z = x + y + z;;
let add (x, y, z) = x + y + z;;

curry3 add 1 2 3;;
curry3L add 1 2 3;;

uncurry3 plus (1, 2, 3);;
uncurry3L plus (1, 2, 3);;

(* Zadanie 3 *)
let mySumProd xs = 
  List.fold_left (fun (right, left) x -> (right + x, left * x)) (0, 1) xs;;

mySumProd [1; 2; 3; 4];;
mySumProd [0; 1; 2; 3];;
mySumProd [-1; -2; -3; -4];;

(* Zadanie 5 *)
(* a) *)
let rec insertionSort pred xs =
  let rec insert pred x xs =
    match xs with 
      [] -> [x]
      | h::t ->
        if pred x h then x::h::t
        else h::insert pred x t
  in
    match xs with
      [] -> []
      | h::t -> insert pred h (insertionSort pred t)
;;

insertionSort (fun x y -> x <= y) [5; 2; 1; 4; 3];;
insertionSort (fun (x, _)(y, _) -> x <= y) [(5, 1); (2, 1); (5, 2); (4, 1); (2, 2)];;
insertionSort (fun (x, _)(y, _) -> x >= y) [(5, 1); (2, 1); (5, 2); (4, 1); (2, 2)];;

(* b) *)
let rec mergeSort pred xs =
  let halve xs =
    let rec halveHelper n ys =
      if n = 0 then ([], ys)
      else match ys with
        [] -> ([], [])
        | h::t ->
          let (l, r) = halveHelper (n - 1) t
          in (h::l, r)
    in halveHelper (List.length xs / 2) xs
  in
    let rec merge pred xs ys =
    match (xs, ys) with
      [], l -> l 
      | l, [] -> l
      | hx::tx, hy::ty ->
        if pred hx hy then hx::merge pred tx ys
        else hy::merge pred xs ty
  in
    match xs with 
      [] -> []
      | [x] -> [x]
      | h::t as list -> 
        let l, r = halve list
        in merge pred (mergeSort pred l) (mergeSort pred r)
;;

mergeSort (fun x y -> x <= y) [5; 2; 1; 4; 3];;
mergeSort (fun (x, _)(y, _) -> x <= y) [(5, 1); (2, 1); (5, 2); (4, 1); (2, 2)];;
mergeSort (fun (x, _)(y, _) -> x >= y) [(5, 1); (2, 1); (5, 2); (4, 1); (2, 2)];;
